### PR TITLE
TINKERPOP-1467 Corrected a number of problems in close() operations for the driver [tp31]

### PR DESCRIPTION
**Pull request by spmallette**
https://issues.apache.org/jira/browse/TINKERPOP-1467

This was more of a commit than I wanted for `tp31`, but `close()` was really messed up. Fixed a number of race conditions and other logic that would allow the driver to hang on close. Also made it so that the `Cluster` makes an attempt to clean up any `Client` instances that it spawns.

Tested with `mvn clean install` and endlessly with `mvn verify -pl gremlin-server -DskipIntegrationTests=false` (basically ran it for a whole day over and over again).

VOTE +1

### CODE REVIEW

File: gremlin-server/src/test/java/org/apache/tinkerpop/gremlin/server/GremlinDriverIntegrateTest.java [hunk 1]
```diff
@@ -1248,6 +1250,73 @@ public void shouldProcessSessionRequestsInOrderAfterTimeout() throws Exception {
...
+        assertEquals(2, session.submit("1+1").all().get().get(0).getInt());
+        assertEquals(2, sessionlessTwo.submit("1+1").all().get().get(0).getInt());
+        assertEquals(2, sessionlessThree.submit("1+1").all().get().get(0).getInt());
+        // dont' send anything on the 4th client
+
+        // close one of these Clients before the Cluster
+        sessionlessThree.close();
+        cluster.close();
+
+        try {
```
Comment (dkuppitz) [inline gremlin-server/src/test/java/org/apache/tinkerpop/gremlin/server/GremlinDriverIntegrateTest.java]: It is not verified, that exceptions are actually thrown. Should be more like:

```
try {
    sessionXyz.submit("1+1").all().get();
    assertTrue(false);
} catch (Exception ex) {
    ....
```

File: docs/src/upgrade/release-3.1.x-incubating.asciidoc [hunk 2]
```diff
@@ -27,6 +27,23 @@ TinkerPop 3.1.5
...
+Please see the link:https://github.com/apache/tinkerpop/blob/3.1.4/CHANGELOG.asciidoc#tinkerpop-315-release-date-XXXXXXXXXXXX[changelog] for a complete list of all the modifications that are part of this release.
+
+Upgrading for Users
+~~~~~~~~~~~~~~~~~~~
+
+Java Driver and close()
+^^^^^^^^^^^^^^^^^^^^^^^
+
+There were a few problems noted around the `close()` of `Cluster` and `Client` instances, including issues that
+presented as system hangs. These issues have been resolved, however, it is worth nothing that an unchecked exception
```
Comment (dkuppitz) [inline docs/src/upgrade/release-3.1.x-incubating.asciidoc]: worth nothing => worth noting

**Comment by dkuppitz** [comment]
As said in a private chat, integration tests were stuck when I ran them first. However, the second attempt succeeded. More investigation is apparently needed, but for this PR:

VOTE: +1

**Comment by spmallette** [comment]
I ran the docker build multiple times today following the report of stuck integration tests. I never ran into the problem, but the adjustments I forced pushed prior to those tests hopefully fixed it.

**Comment by dkuppitz** [comment]
FYI:  3x BUILD SUCCESS in a row.

My vote still holds true.

**Comment by okram** [comment]
VOTE +1.

**Pull request merged**